From 41e56e64d885ac5ec914257137f5e5e8db26ba08 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 21 Aug 2025 09:11:53 +0800
Subject: [PATCH 209/254] mtk: hostapd: properly call ap_sta_remove_link_sta()
 before ap_free_sta()

If the function ap_free_sta() is called only for a single link of a
STA MLD, especially the setup link, it might cause a use-after-free
issue.

The function ap_sta_remove_link_sta should be called before
ap_free_sta() to prevent the issue.

1. add #ifdef CONFIG_IEEE80211BE
2. do not remove other sta_info if reconfiguration is on-going.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/sta_info.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index acd3b8748..10338e1de 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -545,6 +545,11 @@ void hostapd_free_stas(struct hostapd_data *hapd)
 		sta = sta->next;
 		wpa_printf(MSG_DEBUG, "Removing station " MACSTR,
 			   MAC2STR(prev->addr));
+#ifdef CONFIG_IEEE80211BE
+		if (ap_sta_is_mld(hapd, prev) && !(hapd->mld &&
+		    (hapd->mld->removed_links & BIT(hapd->mld_link_id))))
+			ap_sta_remove_link_sta(hapd, prev);
+#endif /* CONFIG_IEEE80211BE */
 		ap_free_sta(hapd, prev);
 	}
 }
@@ -800,6 +805,8 @@ static void ap_handle_session_timer(void *eloop_ctx, void *timeout_ctx)
 		       "session timeout");
 	sta->acct_terminate_cause =
 		RADIUS_ACCT_TERMINATE_CAUSE_SESSION_TIMEOUT;
+	if (ap_sta_is_mld(hapd, sta))
+		ap_sta_remove_link_sta(hapd, sta);
 	ap_free_sta(hapd, sta);
 }
 
@@ -874,8 +881,11 @@ static void ap_sta_assoc_timeout(void *eloop_ctx, void *timeout_ctx)
 	if (sta->flags & WLAN_STA_AUTH)
 		ap_sta_deauthenticate(hapd, sta,
 				      WLAN_REASON_PREV_AUTH_NOT_VALID);
-	else
+	else {
+		if (ap_sta_is_mld(hapd, sta))
+			ap_sta_remove_link_sta(hapd, sta);
 		ap_free_sta(hapd, sta);
+	}
 }
 
 
-- 
2.45.2

